lib/heuristics/periodic_heuristic.rb

```diff
@@ -1010,23 +993,64 @@ def get_unassigned_info(custom_id, service_in_vrp, reason)
       }
     end
 
-    def insert_point_in_route(route_data, point_to_add, first_visit = true)
+    def deduce_number_in_sequence(day, id)
+      return [1, 1] unless @services_data[id][:used_days].any?
+
+      @services_data[id][:used_days].sort!
+      @services_data[id][:assigned_visits].sort!
+      day_before = @services_data[id][:used_days].reverse.find{ |d| d < day }
+      current_index = day_before ? @services_data[id][:assigned_visits][@services_data[id][:used_days].find_index(day_before)] + 1 : 1
+
+      if day_before && @services_data[id][:raw].maximum_lapse
+        while day - day_before > @services_data[id][:raw].maximum_lapse
+          current_index += 1
+          day -= @services_data[id][:heuristic_period]
+        end
+      end
+
+      newly_assigned_index = current_index
+      if @services_data[id][:assigned_visits].include?(current_index)
+        # This means we need to shift visits number
+        new_indices = @services_data[id][:assigned_visits].collect.with_index{ |visit_number, visit_index|
+          if visit_number >= current_index
+            @services_data[id][:used_vehicles].each{ |vehicle_id|
+              stop_in_route = @candidate_routes[vehicle_id][@services_data[id][:used_days][visit_index]][:stops].find{ |s|
+                s[:id] == id
+              }
+              stop_in_route[:number_in_sequence] = visit_number + 1 if stop_in_route
+            }
+            visit_number + 1
+          else
+            visit_number
+          end
+        }
+        newly_assigned_index = ((new_indices + [current_index]) - @services_data[id][:assigned_visits]).first
+      end
+      [current_index, newly_assigned_index]
+    end
+
+    def insert_point_in_route(route_data, point_to_add, add_same_located_points, nb_in_seq = nil)
```

Comment:
`add_same_located_points` doesn't sounds really expressive on the fact it is a boolean

```suggestion
    def insert_point_in_route(route_data, point_to_add, is_same_location, nb_in_seq = nil)
```

Reply:
What is `nb_in_seq` ?

Reply:
It is number_in_sequence, that is the number of the visit among all visits. Sometimes we know the number of the visit we insert, sometimes we don't.

In fact it would be easier not to keep track of this in the code, but unassigned reasons are more precise if we do. I am not sure if it is worth the work.